src/cargo/ops/cargo_add/mod.rs

```diff
@@ -529,27 +544,87 @@ fn get_latest_dependency(
                 }
             };
-            let latest = possibilities
-                .iter()
-                .max_by_key(|s| {
-                    // Fallback to a pre-release if no official release is available by sorting them as
-                    // less.
-                    let stable = s.version().pre.is_empty();
-                    (stable, s.version())
-                })
-                .ok_or_else(|| {
-                    anyhow::format_err!(
-                        "the crate `{dependency}` could not be found in registry index."
-                    )
-                })?;
+
+            possibilities.sort_by_key(|s| {
+                // Fallback to a pre-release if no official release is available by sorting them as
+                // less.
+                let stable = s.version().pre.is_empty();
+                (stable, s.version().clone())
+            });
+
+            let mut latest = possibilities.last().ok_or_else(|| {
+                anyhow::format_err!(
+                    "the crate `{dependency}` could not be found in registry index."
+                )
+            })?;
+
+            if config.cli_unstable().msrv_policy && honor_rust_version {
+                if let Some((req_major, req_minor, req_patch)) =
+                    spec.rust_version().map(parse_rust_version)
+                {
+                    let rust_versions = possibilities
+                        .iter()
+                        .map(|s| {
+                            let rust_version = s.rust_version().map(parse_rust_version);
+                            (s, rust_version)
+                        })
+                        .collect::<Vec<_>>();
+
+                    // Find the latest version of the dep which has a compatible rust-version,
+                    // treating candidates without a rust-version as compatible.
+                    let (latest_msrv, _) = rust_versions
+                        .iter()
+                        .filter(|(_, v)| {
+                            v.map(|(major, minor, patch)| {
+                                // Determine whether or not the rust-versions are compatible by
+                                // comparing the lowest possible versions each of them could
+                                // represent.
+                                (req_major, req_minor, req_patch.unwrap_or(0))
+                                    >= (major, minor, patch.unwrap_or(0))
+                            })
+                            .unwrap_or(true)
+                        })
+                        .last()
+                        .ok_or_else(|| {
+                            let mut error_msg = format!(
+                                "could not find version of crate `{dependency}` that satisfies \
+                                 this package's rust-version\n\
+                                 help: use `--ignore-rust-version` to override this behavior"
```

Comment:
I don't love the lack of styling here, but as far as I can tell there isn't an easier way to emit complicated messages like this in cargo like there is in rustc. There's a couple places in cargo that do this same trick (just putting in a newline after an error), so there is *some* precedence, but I'm also fine just making a slightly longer error message that contains all of the details on a single line.

Reply:
My hope is that we'll be able to adopt `anstream` and then we can put styling wherever we want.

I've not broached the subject yet as I've been wanting to give it more runtime in other places to see if any problems get identified